syntax = "proto3";

package mangarr.tachibridge.extensions;

option go_package = ".";
option java_multiple_files = true;
option java_package = "mangarr.tachibridge.extensions";

// Common data models

message Title {
  string url = 1;
  string title = 2;
  string artist = 3;
  string author = 4;
  string description = 5;
  string genre = 6;
  int32 status = 7;
  string thumbnail_url = 8;
  bool initialized = 9;
}

message Chapter {
  string url = 1;
  string name = 2;
  int64 date_upload = 3;
  float chapter_number = 4;
  string scanlator = 5;
}

message Page {
  int32 index = 1;
  string url = 2;
  string image_url = 3;
}

message Filter {
  string name = 1;
  string type = 2;
  string data = 3;
}

message SourceInfo {
  int64 id = 1;
  string name = 2;
  string lang = 3;
  bool is_nsfw = 4;
  bool supports_latest = 5; // Added supports_latest field
}

message ExtensionInfo {
  string pkg_name = 1;
  string name = 2;
  string version = 3;
  string lang = 4;
  bool nsfw = 5;
  repeated SourceInfo sources = 6;
}

// Source listing

message ListSourcesRequest {}

message ListSourcesResponse {
  repeated SourceInfo sources = 1;
}

// Title search

message SearchTitleRequest {
  int64 source_id = 1;
  string query = 2;
  int32 page = 3;
}

message TitlesPageResponse {
  repeated Title titles = 1;
  bool has_next_page = 2;
}

// Popular titles

message GetPopularTitlesRequest {
  int64 source_id = 1;
  int32 page = 2;
}

// Latest titles

message GetLatestTitlesRequest {
  int64 source_id = 1;
  int32 page = 2;
}

// Title details

message GetTitleDetailsRequest {
  int64 source_id = 1;
  string title_url = 2;
}

message TitleResponse {
  Title title = 1;
}

// Chapter list

message GetChaptersListRequest {
  int64 source_id = 1;
  string title_url = 2;
}

message ChaptersListResponse {
  repeated Chapter chapters = 1;
}

// Page list

message GetPagesListRequest {
  int64 source_id = 1;
  string chapter_url = 2;
}

message PagesListResponse {
  repeated Page pages = 1;
}

// Filters

message GetFiltersRequest {
  int64 source_id = 1;
}

message FiltersResponse {
  repeated Filter filters = 1;
}

// Preferences

message SetPreferenceRequest {
  int64 source_id = 1;
  string key = 2;
  string value = 3;
}

message SetPreferenceResponse {
  bool success = 1;
  string error = 2;
}

// Extension API (from extension.proto)

message InstallExtensionRequest {
  string package_name = 1;
}

message InstallExtensionResponse {
  bool success = 1;
  ExtensionInfo extension = 2;
  string error = 3;
}

message UninstallExtensionRequest {
  string package_name = 1;
}

message UninstallExtensionResponse {
  bool success = 1;
  string error = 2;
}

message UpdateExtensionRequest {
  string package_name = 1;
}

message ListRepoExtensionsRequest {}

message ListRepoExtensionsResponse {
  repeated ExtensionInfo extensions = 1;
}
